File: src/ladym/adapter.py
"""Langchain → ladyM bridge layer.

Houses the adapters that wrap host-owned langchain objects (BaseChatModel,
Embeddings) into ladyM's own provider abstractions, plus ``ModelRouting`` —
the typed injection config. langchain types appear only under TYPE_CHECKING
so importing this module needs no langchain at runtime.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING

from .providers.llm import LLMProvider
from .storage.embeddings import EmbeddingProvider

if TYPE_CHECKING:
    from langchain_core.embeddings import Embeddings
    from langchain_core.language_models import BaseChatModel


def _to_lc(msg):
    """Convert a ladyM Message dict to a langchain message (moved from providers.llm)."""
    from langchain_core.messages import AIMessage, HumanMessage, SystemMessage

    if msg["role"] == "system":
        return SystemMessage(content=msg["content"])
    if msg["role"] == "assistant":
        return AIMessage(content=msg["content"])
    return HumanMessage(content=msg["content"])


class LangChainLLMProvider(LLMProvider):
    """Wrap a langchain ``BaseChatModel`` as a ladyM ``LLMProvider`` (moved from providers.llm)."""

    name = "langchain"

    def __init__(self, chat_model, structured_method: str = "function_calling"):
        self._cm = chat_model
        self._sm = structured_method

    def complete(self, messages, **params):
        return self._cm.invoke([_to_lc(m) for m in messages]).content

    def complete_structured(self, messages, schema, **params):
        runner = self._cm.with_structured_output(schema, method=self._sm)
        out = runner.invoke([_to_lc(m) for m in messages])
        return (
            out
            if isinstance(out, dict)
            else (out.model_dump() if hasattr(out, "model_dump") else dict(out))
        )


class LangChainEmbeddingAdapter(EmbeddingProvider):
    """Bridge a langchain ``Embeddings`` into ladyM's ``EmbeddingProvider``.

    ``dim`` starts ``None`` and is probed on the first :meth:`embed` call —
    same pattern as ``OllamaEmbedding``, so Engine's ``_ensure_provider_dim``
    handles it without special-casing.
    """

    def __init__(self, embeddings: Embeddings):
        self._lc = embeddings
        self.dim: int | None = None

    def embed(self, text: str) -> list[float]:
        vec = self._lc.embed_query(text)
        if self.dim is None:
            self.dim = len(vec)
        return vec

    def embed_batch(self, texts: list[str]) -> list[list[float]]:
        return self._lc.embed_documents(texts)


@dataclass
class ModelRouting:
    """Inject host-owned langchain models, bypassing ladyM's own LLM/embedding config.

    Unset fields fall back to Config / heuristic. Field names mirror ``NAMED_OPS``
    so ``getattr(routing, op, None)`` resolves each cognitive operation.
    """

    consolidate: BaseChatModel | None = None
    proceduralize: BaseChatModel | None = None
    attention_gate: BaseChatModel | None = None
    l5_mental_model: BaseChatModel | None = None
    l6_forward_intent: BaseChatModel | None = None
    embedding: Embeddings | None = None

File: adapter/adapter.go
// Package adapter holds the host-model injection bridge.
//
// In the Python port this wrapped langchain ChatModel/Embeddings objects. The
// Go port integrates langchain-golang: hosts inject their ChatModel /
// Embeddings via WrapChatModel / WrapEmbeddings and assign the results to
// ModelRouting. Unset fields fall back to Config / heuristic mode.
package adapter

import (
	"context"

	"github.com/ProjAnvil/LadyM/providers"
	"github.com/ProjAnvil/LadyM/storage"
	"github.com/projanvil/langchain-golang/core/embeddings"
	"github.com/projanvil/langchain-golang/core/language"
)

// WrapChatModel adapts a host-owned langchain-golang ChatModel into an
// LLMProvider suitable for ModelRouting — the Go equivalent of Python's
// ModelRouting fields accepting langchain BaseChatModel objects directly.
// structuredMethod mirrors Python's structured_method ("function_calling" |
// "json_mode" | "json_schema"; "" = provider default).
func WrapChatModel(cm language.ChatModel, structuredMethod string) providers.LLMProvider {
	return providers.WrapChatModel(cm, structuredMethod)
}

// WrapEmbeddings bridges a host-owned langchain-golang Embeddings into
// ladyM's EmbeddingProvider, mirroring Python's LangChainEmbeddingAdapter.
// Dim starts at 0 and is probed on the first Embed call (deferred-dim
// pattern, same as OllamaEmbedding / Python's dim=None).
func WrapEmbeddings(e embeddings.Embeddings) storage.EmbeddingProvider {
	return &langchainEmbedding{lc: e}
}

type langchainEmbedding struct {
	lc  embeddings.Embeddings
	dim int
}

func toFloat32s(v []float64) []float32 {
	out := make([]float32, len(v))
	for i, x := range v {
		out[i] = float32(x)
	}
	return out
}

func (a *langchainEmbedding) Embed(text string) ([]float32, error) {
	vec, err := a.lc.EmbedQuery(context.Background(), text)
	if err != nil {
		return nil, err
	}
	if a.dim == 0 {
		a.dim = len(vec)
	}
	return toFloat32s(vec), nil
}

func (a *langchainEmbedding) EmbedBatch(texts []string) ([][]float32, error) {
	vecs, err := a.lc.EmbedDocuments(context.Background(), texts)
	if err != nil {
		return nil, err
	}
	out := make([][]float32, len(vecs))
	for i, v := range vecs {
		if a.dim == 0 {
			a.dim = len(v)
		}
		out[i] = toFloat32s(v)
	}
	return out, nil
}

func (a *langchainEmbedding) Dim() int { return a.dim }

func (a *langchainEmbedding) HealthCheck() (bool, string) {
	if _, err := a.lc.EmbedQuery(context.Background(), "ping"); err != nil {
		return false, err.Error()
	}
	return true, "ok"
}

// ModelRouting injects host-owned models, bypassing LadyM's own config. Unset
// fields fall back to Config / heuristic. Field names mirror providers.NAMED_OPS.
type ModelRouting struct {
	Embedding       storage.EmbeddingProvider
	Consolidate     providers.LLMProvider
	Proceduralize   providers.LLMProvider
	AttentionGate   providers.LLMProvider
	L5MentalModel   providers.LLMProvider
	L6ForwardIntent providers.LLMProvider
}

// Get returns the injected LLM provider for op, or nil.
func (r *ModelRouting) Get(op string) providers.LLMProvider {
	if r == nil {
		return nil
	}
	switch op {
	case "consolidate":
		return r.Consolidate
	case "proceduralize":
		return r.Proceduralize
	case "attention_gate":
		return r.AttentionGate
	case "l5_mental_model":
		return r.L5MentalModel
	case "l6_forward_intent":
		return r.L6ForwardIntent
	default:
		return nil
	}
}
